lib/Parse/ParseType.cpp

```diff
@@ -981,14 +982,19 @@ ParserResult<TupleTypeRepr> Parser::parseTypeTupleBody() {
         diag.fixItReplace(SourceRange(element.NameLoc), "_");
     }
 
-    if (element.NameLoc.isValid() || element.SecondNameLoc.isValid()) {
+    if (element.SecondNameLoc.isValid()) {
       // Form the named parameter type representation.
+      element.UnderscoreLoc = element.NameLoc;
       element.Name = element.SecondName;
       element.NameLoc = element.SecondNameLoc;
-      element.UnderscoreLoc = element.NameLoc;
     }
   }
 
+  if (isFunctionType)
+    SyntaxContext->setTransparent();
```

Comment:
Could you clarify why do we need this? If a `FunctionType` node is already created, the parent context will do nothing if the `FunctionType` is the only node. `setTransparent` will disallow the creation of an `UnknownType` wrapper if there are other nodes; is this why we need it here?

Reply:
The structure here is something like this:
```swift
func parseTypeTupleBody() {
   SyntaxContextRAII(SyntaxContext)
   consumeToken(tok::l_paren) // '('
   parseTupleTypeElementList() // tuple-type-element-list
   consumeToken(tok::r_paren) // ')'
   
   if isFunctionType {
    // propagate to the parent stack.
    SyntaxContext.setTransparent()
   } else {
    //  '(' tuple-type-element-list ')'
    SyntaxContext.setCreate(TupleType)
   }
}

func parseType() {
   SyntaxContextRAII(SyntaxContext)
   parseTypeTupleBody()
   
   if (isFunctionType) {
    // '(' tuple-type-element-list ')' 'throws'? '->' type
    Context.setCreateType(FunctionType)
   } else {
    // type | unknown-type 
    Context.setCoerce(.type)
   }
}
```
If we see `->` or `throws` after `r_paren`, we have to propagate `'(' tuple-type-element-list ')'` part back to `parseType()`

Reply:
I see. Thank you for explaining! could we add some comments there to keep track of this scope-misalignment?

Reply:
Thinking a little big more about this, how about we make the first three children of a `FunctionType` a `TupleType`?  This allows us to get rid of `setTransparent` and to simplify the structure of `FunctionType`.

Reply:
Hmm, although that's convenient solution for now, I'd like to avoid that because it's against @CodaFi's efforts of re-modeling function type parameters.
How about to parse just `TypeTuple` in `parseTypeTupleBody`, then decompose it to `'(' TupleTypeElementList ')'` and repack into `FunctionType` in `parseType`?

Reply:
that'll add another operation (decompose) to SyntaxParsingContext, right? If changing function type's structure isn't a good option, i'd say what you have for now is good enough.

Reply:
> that'll add another operation (decompose) to SyntaxParsingContext, right?

Not really. We can `SyntaxParsingContext::popIf()`, build `FunctionType` with `FunctionTypeSyntaxBuilder`, then `SyntaxParsingContext::addSyntax()`.